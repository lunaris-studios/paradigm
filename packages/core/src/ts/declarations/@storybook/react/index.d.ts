import * as Abstract from "@paradigmjs/abstract";
import * as Addons from "@storybook/addons";
import * as SR from "@storybook/react";

import "@storybook/react";

declare module "@storybook/react" {
	/**
	 * Metadata to configure the stories for an abstract component.
	 */
	declare type StorybookComponent =
		| React.ComponentType
		| Abstract.AbstractComponent
		| Abstract.AbstractPureComponent;
	export declare type Meta<
		TComponent extends StorybookComponent = any,
		TArgs = Addons.DefaultArgs
	> = Addons.BaseMeta<TComponent> &
		Addons.Annotations<Component.Spinner, React.ReactElement<unknown>>;
}
