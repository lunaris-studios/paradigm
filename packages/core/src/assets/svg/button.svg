<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 258 100">
  <defs>
    <linearGradient id="New_Gradient_Swatch_33" x1="129.4" x2="129.4" y1="81.6" y2="8.4" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#651928"/>
      <stop offset="1" stop-color="#773642"/>
    </linearGradient>
    <linearGradient id="New_Gradient_Swatch_67" x1="166.5" x2="121.3" y1="127.4" y2="27.1" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#773642"/>
      <stop offset="1" stop-color="#e45053"/>
    </linearGradient>
    <linearGradient id="New_Gradient_Swatch_68" x1="108.3" x2="135.4" y1="-19" y2="63.3" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#fe7c79"/>
      <stop offset="1" stop-color="#e45053" stop-opacity="0"/>
    </linearGradient>
    <linearGradient id="linear-gradient" x1="129" x2="129" y1="78" y2="10.1" gradientTransform="translate(-1 -2)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#fe7c79" stop-opacity="0"/>
      <stop offset="1" stop-color="#fe7c79" stop-opacity=".7"/>
    </linearGradient>
    <linearGradient id="linear-gradient-2" x1="129.4" x2="129.4" y1="68.6" y2="23.4" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#fe7c79" stop-opacity="0"/>
      <stop offset="1" stop-color="#fe7c79"/>
    </linearGradient>
    <linearGradient id="linear-gradient-3" x1="129.4" x2="129.4" y1="67.6" y2="22.4" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#963a4c" stop-opacity="0"/>
      <stop offset="1" stop-color="#963a4c"/>
    </linearGradient>
    <linearGradient id="New_Gradient_Swatch_12" x1="182.7" x2="182.7" y1="78.9" y2="66.8" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#e45053" stop-opacity="0"/>
      <stop offset="1" stop-color="#e45053"/>
    </linearGradient>
    <linearGradient id="New_Gradient_Swatch_12-2" x1="85.8" x2="85.8" y1="80.1" y2="61" xlink:href="#New_Gradient_Swatch_12"/>
    <linearGradient id="New_Gradient_Swatch_12-3" x1="147.9" x2="147.9" y1="78.9" y2="67.5" xlink:href="#New_Gradient_Swatch_12"/>
    <linearGradient id="linear-gradient-4" x1="232.1" x2="232.1" y1="79.6" y2="9.4" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#4a4151"/>
      <stop offset=".5" stop-color="#695a6a"/>
      <stop offset="1" stop-color="#97868a"/>
    </linearGradient>
    <linearGradient id="linear-gradient-5" x1="214.4" x2="222.8" y1="31.5" y2="39.9" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#ab9da4" stop-opacity="0"/>
      <stop offset=".5" stop-color="#ab9da4"/>
      <stop offset="1" stop-color="#ab9da4" stop-opacity="0"/>
    </linearGradient>
    <linearGradient id="linear-gradient-6" x1="232.1" x2="232.1" y1="79.6" y2="9.4" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#ab9da4" stop-opacity="0"/>
      <stop offset="1" stop-color="#ab9da4"/>
    </linearGradient>
    <linearGradient id="linear-gradient-7" x1="147" x2="147" y1="-192.9" y2="-197.9" gradientTransform="rotate(-61 409 -144)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#695a6a"/>
      <stop offset="1" stop-color="#97868a"/>
    </linearGradient>
    <linearGradient id="linear-gradient-8" x1="220" x2="233" y1="44" y2="44" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#322838"/>
      <stop offset=".1" stop-color="#322838" stop-opacity=".9"/>
      <stop offset=".4" stop-color="#322838" stop-opacity=".4"/>
      <stop offset=".8" stop-color="#322838" stop-opacity=".1"/>
      <stop offset="1" stop-color="#322838" stop-opacity="0"/>
    </linearGradient>
    <linearGradient id="New_Gradient_Swatch_10" x1="529.7" x2="547.7" y1="44.5" y2="44.5" gradientTransform="matrix(-1 0 0 1 745 0)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#773642"/>
      <stop offset=".2" stop-color="#773642" stop-opacity=".7"/>
      <stop offset=".4" stop-color="#773642" stop-opacity=".4"/>
      <stop offset=".7" stop-color="#773642" stop-opacity=".2"/>
      <stop offset=".8" stop-color="#773642" stop-opacity=".1"/>
      <stop offset="1" stop-color="#773642" stop-opacity="0"/>
    </linearGradient>
    <linearGradient id="linear-gradient-9" x1="286.6" x2="286.6" y1="79.6" y2="9.4" gradientTransform="matrix(-1 0 0 1 314 0)" xlink:href="#linear-gradient-4"/>
    <linearGradient id="linear-gradient-10" x1="268.9" x2="277.3" y1="31.5" y2="39.9" gradientTransform="matrix(-1 0 0 1 314 0)" xlink:href="#linear-gradient-5"/>
    <linearGradient id="linear-gradient-11" x1="286.6" x2="286.6" y1="79.6" y2="9.4" gradientTransform="matrix(-1 0 0 1 314 0)" xlink:href="#linear-gradient-6"/>
    <linearGradient id="linear-gradient-12" x1="173.1" x2="173.1" y1="-145.1" y2="-150.1" gradientTransform="scale(-1 1) rotate(-61 251 121)" xlink:href="#linear-gradient-7"/>
    <linearGradient id="linear-gradient-13" x1="275.5" x2="288.5" gradientTransform="matrix(-1 0 0 1 314 0)" xlink:href="#linear-gradient-8"/>
    <linearGradient id="New_Gradient_Swatch_10-2" x1="475.8" x2="493.8" y1="44.5" y2="44.5" gradientTransform="translate(-431)" xlink:href="#New_Gradient_Swatch_10"/>
    <linearGradient id="linear-gradient-14" x1="218" x2="240.3" y1="45" y2="45" gradientUnits="userSpaceOnUse">
      <stop offset=".2" stop-color="#594858" stop-opacity="0"/>
      <stop offset=".5" stop-color="#7d6c75" stop-opacity=".6"/>
      <stop offset=".7" stop-color="#97868a"/>
    </linearGradient>
    <linearGradient id="linear-gradient-15" x1="229.1" x2="229.1" y1="82" y2="8" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#4a4151"/>
      <stop offset=".3" stop-color="#594858" stop-opacity="0"/>
      <stop offset=".3" stop-color="#5b4a59" stop-opacity="0"/>
      <stop offset=".5" stop-color="#6d5c68" stop-opacity=".3"/>
      <stop offset=".6" stop-color="#7c6b74" stop-opacity=".6"/>
      <stop offset=".7" stop-color="#88777e" stop-opacity=".8"/>
      <stop offset=".8" stop-color="#907f85" stop-opacity=".9"/>
      <stop offset=".9" stop-color="#958489"/>
      <stop offset="1" stop-color="#97868a"/>
    </linearGradient>
    <linearGradient id="linear-gradient-16" x1="229.1" x2="229.1" y1="82" y2="8" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#594858" stop-opacity="0"/>
      <stop offset=".2" stop-color="#7d6c75" stop-opacity=".6"/>
      <stop offset=".3" stop-color="#97868a"/>
      <stop offset=".4" stop-color="#594858" stop-opacity="0"/>
    </linearGradient>
    <linearGradient id="linear-gradient-17" x1="91.8" x2="113.8" y1="45" y2="45" gradientTransform="matrix(-1 0 0 1 134 0)" xlink:href="#linear-gradient-14"/>
    <linearGradient id="linear-gradient-18" x1="102.8" x2="102.8" y1="82" gradientTransform="matrix(-1 0 0 1 134 0)" xlink:href="#linear-gradient-15"/>
    <linearGradient id="linear-gradient-19" x1="102.8" x2="102.8" y1="82" gradientTransform="matrix(-1 0 0 1 134 0)" xlink:href="#linear-gradient-16"/>
    <linearGradient id="New_Gradient_Swatch_33-2" x1="129.9" x2="129.9" y1="55.8" y2="33.3" gradientTransform="translate(0 2)" xlink:href="#New_Gradient_Swatch_33"/>
    <linearGradient id="New_Gradient_Swatch_33-6" x1="129.9" x2="129.9" y1="55.8" y2="33.3" xlink:href="#New_Gradient_Swatch_33"/>
    <linearGradient id="New_Gradient_Swatch_27" x1="129.9" x2="129.9" y1="50.4" y2="27.9" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#beadb6"/>
      <stop offset=".5" stop-color="#beadb6" stop-opacity="0"/>
    </linearGradient>
    <style>
      .cls-2{fill:#651928}.cls-22{mix-blend-mode:overlay}.cls-16{fill:#963a4c}.cls-22{opacity:.7}.cls-16,.cls-27{mix-blend-mode:darken}.cls-19,.cls-36,.cls-37,.cls-41,.cls-46,.cls-47{fill:none}.cls-19,.cls-36,.cls-41{stroke:#322838}.cls-19{stroke-miterlimit:10}.cls-19,.cls-37,.cls-46,.cls-47{stroke-width:4px}.cls-24{fill:#695a6a}.cls-26{fill:#594858}.cls-27{fill:#322838;opacity:.5}.cls-36,.cls-37,.cls-41,.cls-46,.cls-47{stroke-linecap:round;stroke-linejoin:round}.cls-36{stroke-width:8px}.cls-37{stroke:#594858}.cls-41{stroke-width:2px}.cls-46{stroke:url(#New_Gradient_Swatch_33-2)}.cls-47{stroke:url(#New_Gradient_Swatch_33-6)}.cls-49{fill:url(#New_Gradient_Swatch_27)}
    </style>
  </defs>
  <g style="isolation:isolate">
    <path d="M228 12a16 16 0 0116 16v38a16 16 0 01-16 16H31a16 16 0 01-16-16V28a16 16 0 0116-16h197m0-2H31a18 18 0 00-18 18v38a18 18 0 0018 18h197a18 18 0 0018-18V28a18 18 0 00-18-18z" class="cls-2"/>
    <rect width="228.8" height="69.3" x="15" y="12" class="cls-2" rx="16"/>
    <rect width="228.8" height="69.3" x="15" y="12" class="cls-2" rx="16"/>
    <rect width="216.8" height="57.3" x="20" y="16" class="cls-2" rx="10"/>
    <rect width="216.8" height="57.3" x="21" y="17" class="cls-2" rx="10"/>
    <rect width="216.8" height="57.3" x="21" y="17" fill="#651928" opacity=".2" rx="10"/>
    <path d="M228 26a3 3 0 013 3v38a3 3 0 01-3 3H31a3 3 0 01-3-3V29a3 3 0 013-3h197m0-1H31a4 4 0 00-4 4v38a4 4 0 004 4h197a4 4 0 004-4V29a4 4 0 00-4-4z" class="cls-2"/>
    <path d="M228 25a3 3 0 013 3v38a3 3 0 01-3 3H31a3 3 0 01-3-3V28a3 3 0 013-3h197m0-1H31a4 4 0 00-4 4v38a4 4 0 004 4h197a4 4 0 004-4V28a4 4 0 00-4-4z" class="cls-2"/>
    <path fill="url(#New_Gradient_Swatch_33)" d="M228 10a16 16 0 0116 16v38a16 16 0 01-16 16H31a16 16 0 01-16-16V26a16 16 0 0116-16h197m0-2H31a18 18 0 00-18 18v38a18 18 0 0018 18h197a18 18 0 0018-18V26a18 18 0 00-18-18z"/>
    <rect width="228.8" height="69.3" x="15" y="10" fill="url(#New_Gradient_Swatch_67)" rx="16"/>
    <rect width="228.8" height="69.3" x="15" y="10" fill="url(#New_Gradient_Swatch_68)" rx="16"/>
    <rect width="216.8" height="57.3" x="20" y="14" fill="url(#linear-gradient)" rx="10"/>
    <rect width="216.8" height="57.3" x="21" y="15" fill="#e45053" rx="10"/>
    <rect width="216.8" height="57.3" x="21" y="15" fill="#963a4c" opacity=".2" rx="10"/>
    <path fill="url(#linear-gradient-2)" d="M228 24a3 3 0 013 3v38a3 3 0 01-3 3H31a3 3 0 01-3-3V27a3 3 0 013-3h197m0-1H31a4 4 0 00-4 4v38a4 4 0 004 4h197a4 4 0 004-4V27a4 4 0 00-4-4z" opacity=".3"/>
    <path fill="url(#linear-gradient-3)" d="M228 23a3 3 0 013 3v38a3 3 0 01-3 3H31a3 3 0 01-3-3V26a3 3 0 013-3h197m0-1H31a4 4 0 00-4 4v38a4 4 0 004 4h197a4 4 0 004-4V26a4 4 0 00-4-4z" opacity=".7"/>
    <path fill="url(#New_Gradient_Swatch_12)" d="M185 67l-5 5 5 7V67z"/>
    <path d="M185 67l-3 5 3 5V67z" class="cls-16"/>
    <path fill="url(#New_Gradient_Swatch_12-2)" d="M85 61l1 10-4 8 6 1 1-8-4-11z"/>
    <path d="M89 72l-2-5v5l-2 8 3 2 1-7v-3z" class="cls-16"/>
    <path fill="url(#New_Gradient_Swatch_12-3)" d="M146 67l4 5-3 7-1-12z"/>
    <path d="M146 67l2 5-2 4v-9z" class="cls-16"/>
    <path d="M232 11h-15v71h15a16 16 0 0016-16V27a16 16 0 00-16-16z" class="cls-19"/>
    <path d="M232 9h-15v71h15a16 16 0 0016-16V25a16 16 0 00-16-16z" class="cls-19"/>
    <path fill="url(#linear-gradient-4)" d="M232 9h-15v71h15a16 16 0 0016-16V25a16 16 0 00-16-16z"/>
    <path fill="url(#linear-gradient-5)" d="M232 9h-15v71h15a16 16 0 0016-16V25a16 16 0 00-16-16z"/>
    <path fill="url(#linear-gradient-6)" d="M222 75V14h10a11 11 0 0111 11v39a11 11 0 01-11 11z"/>
    <path d="M223 74V15h9a10 10 0 0110 10v39a10 10 0 01-10 10z" class="cls-24"/>
    <path fill="url(#linear-gradient-7)" d="M237 61l4 4h6l-10-4z"/>
    <path d="M237 61l5-1 3 4-8-3z" class="cls-26"/>
    <path d="M237 61l5-1 3 4-8-3z" class="cls-27"/>
    <path fill="url(#linear-gradient-8)" d="M220 8h13v72h-13z" opacity=".5" style="mix-blend-mode:darken"/>
    <path fill="url(#New_Gradient_Swatch_10)" fill-opacity=".5" d="M197 8h18v73h-18z" style="mix-blend-mode:darken"/>
    <path d="M44 11H28a16 16 0 00-16 16v39a16 16 0 0016 16h16z" class="cls-19"/>
    <path d="M44 9H28a16 16 0 00-16 16v39a16 16 0 0016 16h16z" class="cls-19"/>
    <path fill="url(#linear-gradient-9)" d="M44 9H28a16 16 0 00-16 16v39a16 16 0 0016 16h16z"/>
    <path fill="url(#linear-gradient-10)" d="M44 9H28a16 16 0 00-16 16v39a16 16 0 0016 16h16z"/>
    <path fill="url(#linear-gradient-11)" d="M28 76a12 12 0 01-12-12V25a12 12 0 0112-12h12v63z"/>
    <path d="M28 74a10 10 0 01-10-10V25a10 10 0 0110-10h10v59z" class="cls-24"/>
    <path fill="url(#linear-gradient-12)" d="M23 61l-4 4h-6l10-4z"/>
    <path d="M23 61l-5-1-3 4 8-3z" class="cls-26"/>
    <path d="M23 61l-5-1-3 4 8-3z" class="cls-27"/>
    <path fill="url(#linear-gradient-13)" d="M26 8h13v72H26z" opacity=".5" style="mix-blend-mode:darken"/>
    <path fill="url(#New_Gradient_Swatch_10-2)" fill-opacity=".5" d="M45 8h18v73H45z" style="mix-blend-mode:darken"/>
    <path d="M219 11v24l15 18 5-5-5-5-15 18v22" class="cls-36"/>
    <path d="M219 9v24l15 18 5-5-5-5-15 18v22" class="cls-36"/>
    <path d="M219 9v24l15 18 5-5-5-5-15 18v22" class="cls-37"/>
    <path fill="none" stroke="url(#linear-gradient-14)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M219 9v24l15 18 5-5-5-5-15 18v22"/>
    <path fill="none" stroke="url(#linear-gradient-15)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M219 9v24l15 18 5-5-5-5-15 18v22"/>
    <path fill="none" stroke="url(#linear-gradient-16)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M219 9v24l15 18 5-5-5-5-15 18v22"/>
    <path d="M231 40l-9 12M234 45l-10 14" class="cls-41"/>
    <path d="M41 11v24L26 53l-5-5 5-5 15 18v22" class="cls-36"/>
    <path d="M41 9v24L26 51l-5-5 5-5 15 18v22" class="cls-36"/>
    <path d="M41 9v24L26 51l-5-5 5-5 15 18v22" class="cls-37"/>
    <path fill="none" stroke="url(#linear-gradient-17)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M41 9v24L26 51l-5-5 5-5 15 18v22"/>
    <path fill="none" stroke="url(#linear-gradient-18)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M41 9v24L26 51l-5-5 5-5 15 18v22"/>
    <path fill="none" stroke="url(#linear-gradient-19)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M41 9v24L26 51l-5-5 5-5 15 18v22"/>
    <path d="M29 40l10 12M26 45l10 14" class="cls-41"/>
    <path d="M97 55l-2-2V38l2-2h11l2 2v15l-2 2h-3l3 3h-6v-3zm2-4h7V40h-7zM116 53V38l-1-2h7l-2 2 1 13h6V38l-1-2h6l-1 2v15l-2 2h-11zM139 53l-1-15-1-2h7l-1 2v15l1 2h-7zM155 53V39h-4l-2 2v-6h16v5l-2-1h-4v14l1 2h-6z" class="cls-46"/>
    <path d="M97 53l-2-2V36l2-2h11l2 2v15l-2 2h-3l3 3h-6v-3zm2-4h7V38h-7zM116 51V36l-1-2h7l-2 2 1 13h6V36l-1-2h6l-1 2v15l-2 2h-11zM139 51l-1-15-1-2h7l-1 2v15l1 2h-7zM155 51V37h-4l-2 2v-6h16v5l-2-1h-4v14l1 2h-6z" class="cls-47"/>
    <!-- <text fill="#ffd59f" font-family="KVC-Midgard-Regular,KVC-Midgard" font-size="28" transform="translate(93 53)">
      Quit
    </text> -->
    <path d="M97 53l-2-2V36l2-2h11l2 2v15l-2 2h-3l3 3h-6v-3zm2-4h7V38h-7zM116 51V36l-1-2h7l-2 2 1 13h6V36l-1-2h6l-1 2v15l-2 2h-11zM139 51l-1-15-1-2h7l-1 2v15l1 2h-7zM155 51V37h-4l-2 2v-6h16v5l-2-1h-4v14l1 2h-6z" class="cls-49"/>
  </g>
</svg>
