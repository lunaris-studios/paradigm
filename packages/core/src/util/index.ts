export * from "./_can-use-dom";
export * from "./_compose";
export * from "./_function-utils";
export * from "./_is-key-board-click";
export * from "./_is-node-env";
export * from "./_js-utils";
export * from "./_pipe";
export * from "./_react-utils";
export * from "./_remove-non-html-props";
export * from "./_test-util";

import clamp from "lodash/clamp";

export { clamp };
export { default as isNull } from "lodash/isNull";
export { default as camelCase } from "lodash/camelCase";
