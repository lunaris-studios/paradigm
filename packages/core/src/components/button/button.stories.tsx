import * as Abstract from "@paradigmjs/abstract";
import * as React from "react";
import * as SR from "@storybook/react";

import * as Component from "./button";

const Template: SR.Story<Component.IButtonProps> = (args) => (
	<Component.Button {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};

export default {
	title: "Button",
	component: Component.Button,
} as SR.Meta;
