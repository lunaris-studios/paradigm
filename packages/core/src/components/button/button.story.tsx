import * as React from "react";
import * as Knobs from "@storybook/addon-knobs";
import * as Actions from "@storybook/addon-actions";

import { TestButton } from "./button";

export default {
	title: "Button",
	component: TestButton,
	decorators: [Knobs.withKnobs],
};

export const interactive = () => {
	const label = Knobs.text("Label", "Quit");
	return <TestButton onClick={Actions.action("TestButton")} text={label} />;
};
