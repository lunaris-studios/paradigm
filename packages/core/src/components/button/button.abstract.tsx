import * as Abstract from "@paradigmjs/abstract";
import * as Icons from "@paradigmjs/icons";
import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as Spring from "react-spring";

import * as Common from "~/common";
import * as Components from "~/components";
import * as Util from "~/util";

import * as Styled from "./button.styled";

export interface IAbstractButtonProps extends Common.IActionProps {
	/**
	 * If set to `true`, the button will display in an active state.
	 * This is equivalent to setting `className={Classes.ACTIVE}`.
	 * @default false
	 */
	active: boolean;

	/**
	 * Text alignment within button. By default, icons and text will be centered
	 * within the button. Passing `"left"` or `"right"` will align the button
	 * text to that side and push `icon` and `rightIcon` to either edge. Passing
	 * `"center"` will center the text and icons together.
	 * @default Alignment.CENTER
	 */
	alignText: Protocol.Alignment;

	/**
	 * A ref handler that receives the native HTML element backing this component.
	 */
	elementRef?: (ref: HTMLElement | null) => any;

	/**
	 * Whether this button should expand to fill its container.
	 * @default false
	 */
	fill: boolean;

	/**
	 * Name of a Paradigm UI icon (or an icon element) to render before the text.
	 */
	icon?: Icons.IconName | Common.MaybeElement;

	/**
	 * Whether this button should use large styles.
	 * @default false
	 */
	large: boolean;

	/**
	 * If set to `true`, the button will display a centered loading spinner instead of its contents.
	 * The width of the button is not affected by the value of this prop.
	 * @default false
	 */
	loading: boolean;

	/**
	 * Whether this button should use minimal styles.
	 * @default false
	 */
	minimal: boolean;

	/**
	 * Whether this button should use outlined styles.
	 * @default false
	 */
	outlined: boolean;

	/**
	 * Name of a Paradigm UI icon (or an icon element) to render after the text.
	 */
	rightIcon?: Icons.IconName | Common.MaybeElement;

	/**
	 * Whether this button should use small styles.
	 * @default false
	 */
	small: boolean;

	/**
	 * HTML `type` attribute of button. Accepted values are `"button"`, `"submit"`, and `"reset"`.
	 * Note that this prop has no effect on `AnchorButton`; it only affects `Button`.
	 * @default "button"
	 */
	type: "submit" | "reset" | "button";
}

export interface IAbstractButtonState {
	isActive: boolean;
}

const defaultProps = Object.freeze<IAbstractButtonProps>({
	active: false,
	alignText: Protocol.Alignment.CENTER,
	disabled: false,
	fill: false,
	intent: Protocol.Intent.NONE,
	large: false,
	loading: false,
	minimal: false,
	outlined: false,
	small: false,
	type: "button",
});

export const defaultState = Object.freeze<IAbstractButtonState>({
	isActive: false,
});

export abstract class AbstractButton<
	TAbstractButtonAttributes extends Spring.AnimatedProps<React.HTMLAttributes<any>>
> extends Abstract.AbstractPureComponent<
	IAbstractButtonProps & TAbstractButtonAttributes,
	IAbstractButtonState
> {
	constructor(props: IAbstractButtonProps & TAbstractButtonAttributes) {
		super(props);

		this.handleKeyDown = this.handleKeyDown.bind(this);
		this.handleKeyUp = this.handleKeyUp.bind(this);
	}

	public static readonly defaultProps: IAbstractButtonProps = defaultProps;
	public static readonly defaultState: IAbstractButtonState = defaultState;
	public state = defaultState;

	protected buttonRef: HTMLElement | null = null;
	protected refHandlers = {
		button: (ref: HTMLElement | null) => {
			this.buttonRef = ref;
			Util.safeInvoke(this.props.elementRef, ref);
		},
	};

	private currentKeyDown: number = -1;

	public abstract render(): JSX.Element;

	/** */

	public getCommonButtonProps() {
		const { disabled, loading, onClick, tabIndex } = this.props;

		return {
			// ...this.props,
			disabled: disabled || loading,
			onClick: disabled ? undefined : onClick,
			onKeyDown: this.handleKeyDown,
			onKeyUp: this.handleKeyUp,
			ref: this.refHandlers.button,
			tabIndex: disabled ? -1 : tabIndex,
		};
	}

	// we're casting as `any` to get around a somewhat opaque safeInvoke error
	// that "Type argument candidate 'KeyboardEvent<T>' is not a valid type
	// argument because it is not a supertype of candidate
	// 'KeyboardEvent<HTMLElement>'."
	protected handleKeyDown(event: React.KeyboardEvent<any>) {
		if (Util.isKeyboardClick(event.which)) {
			event.preventDefault();
			if (event.which !== this.currentKeyDown) {
				this.setState({ isActive: true });
			}
		}

		this.currentKeyDown = event.which;
		Util.safeInvoke(this.props.onKeyDown, event);
	}

	protected handleKeyUp(event: React.KeyboardEvent<any>) {
		if (!Util.isNull(this.buttonRef) && Util.isKeyboardClick(event.which)) {
			this.setState({ isActive: false });
			// `buttonRef` will always be defined via `getCommonButtonProps()` on the
			// parent implementation.
			this.buttonRef!.click();
		}

		this.currentKeyDown = -1;
		Util.safeInvoke(this.props.onKeyUp, event);
	}

	protected renderChildren(): React.ReactNode {
		const { children, icon, loading, rightIcon, text } = this.props;

		return (
			<React.Fragment>
				{loading && (
					<Components.Spinner key="loading" size={Components.Icon.SIZE_LARGE} />
				)}
				<Components.Icon key="leftIcon" icon={icon} />
				{(!Util.isReactNodeEmpty(text) || !Util.isReactNodeEmpty(children)) && (
					<Styled.Button.Text>
						{text}
						{children}
					</Styled.Button.Text>
				)}
				<Components.Icon key="rightIcon" icon={rightIcon} />
			</React.Fragment>
		);
	}
}
