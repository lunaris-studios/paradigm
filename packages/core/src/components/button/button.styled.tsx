import * as React from "react";
import * as SC from "styled-components";
import * as Spring from "react-spring";
// re-import `styled-components` development mode DOM classnames.
import styled from "styled-components";

import * as Util from "~/util";

/**
 * [Button]
 * - [Button.Container(WRAPPER)]
 * - - [Components.Spinner(COMPONENT)]
 * - - [Components.Icon(COMPONENT)]
 * - - [Button.Text(ELEMENT)]
 */

interface Button {
	Container: Util.SC.Styled<Spring.AnimatedComponent<"button">, IButtonContainerProps>;
	Text: Util.SC.Styled<Spring.AnimatedComponent<"span">, IButtonTextProps>;
}

export const Button = {} as Button;

/**
 * [Button.Container]
 */

interface IButtonContainerProps {}

/**
 * This is overwritten for `AnchorButton` with a corresponding `Spring.animated("a")`
 * via styled-component `as` override.
 */
const AnimatedButtonContainer = Spring.animated("button");

Button.Container = styled(AnimatedButtonContainer)``;

/**
 * [Button.Text]
 */

interface IButtonTextProps {}

/**
 * This is overwritten for `AnchorButton` with a corresponding `Spring.animated("a")`
 * constant.
 */
const AnimatedButtonText = Spring.animated("span");

Button.Text = styled(AnimatedButtonText)``;
