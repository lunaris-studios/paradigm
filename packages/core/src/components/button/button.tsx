import * as React from "react";
import * as Spring from "react-spring";

import * as Common from "~/common";
import * as Util from "~/util";

import * as Abstract from "./button.abstract";
import * as Styled from "./button.styled";

export interface IButtonProps extends Abstract.IAbstractButtonProps {}

export class Button extends Abstract.AbstractButton<
	Spring.AnimatedProps<React.ButtonHTMLAttributes<HTMLButtonElement>>
> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.Button`;

	public render(): JSX.Element {
		return (
			<Styled.Button.Container
				{...Util.removeNonHTMLProps(this.props)}
				{...this.getCommonButtonProps()}
				/** */
				type="button"
			>
				{this.renderChildren()}
			</Styled.Button.Container>
		);
	}
}

export class AnchorButton extends Abstract.AbstractButton<
	Spring.AnimatedProps<React.AnchorHTMLAttributes<HTMLAnchorElement>>
> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.AnchorButton`;

	public render(): JSX.Element {
		const { href, tabIndex = 0 } = this.props;

		const commonProps = this.getCommonButtonProps();

		return (
			<Styled.Button.Container
				{...Util.removeNonHTMLProps(this.props)}
				{...commonProps}
				/** */
				as={Spring.animated.a}
				role="button"
				href={commonProps.disabled ? undefined : href}
				tabIndex={commonProps.disabled ? -1 : tabIndex}
			>
				{this.renderChildren()}
			</Styled.Button.Container>
		);
	}
}
