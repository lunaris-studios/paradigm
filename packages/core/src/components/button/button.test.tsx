import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";

import * as Util from "~/util";

import * as Component from "./button";
import * as Styled from "./button.styled";

// @ts-ignore
import { find } from "styled-components/test-utils";

describe("Button", () => {
	it("renders correctly", () => {
		const node = Util.wrapper(<Component.Button />);
		const ButtonContainer = find(node.baseElement, Styled.Button.Element);
		expect(ButtonContainer).toMatchSnapshot();
	});
});
