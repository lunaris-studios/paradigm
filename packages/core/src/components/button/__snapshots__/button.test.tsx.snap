// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnchorButton renders correctly 1`] = `
<a
  class=""
  role="button"
  tabindex="0"
  type="button"
/>
`;

exports[`Button renders correctly 1`] = `
.c0 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  vertical-align: text-bottom;
}

.c0:not(:empty)::before {
  content: "" !important;
  content: unset !important;
}

.c1 {
  display: block;
}

.c1:not([fill]) {
  fill: currentColor;
}

<button
  class=""
  type="button"
>
  <span
    class="c0"
  >
    <svg
      class="c1"
      data-icon="add"
      height="16"
      viewBox="0 0 16 16"
      width="16"
    >
      <desc
        class=""
      >
        add
      </desc>
      <path
        d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
        fill-rule="evenodd"
      />
    </svg>
  </span>
  <span
    class=""
  >
    button
  </span>
</button>
`;
